core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -73,221 +112,818 @@ public static RecordState forId(byte id) {
+                if (subMap.isEmpty()) {
+                    log.debug("Batch record {} not found for share partition: {}-{}", batch, groupId, topicIdPartition);
+                    throwable = new InvalidRequestException("Batch record not found");
+                    break;
+                }
+
+                // Validate if the request batch has the last offset greater than the last offset of
+                // the last fetched cached batch, then there will be offsets in the request than cannot
+                // be found in the fetched batches.
+                if (i == acknowledgementBatch.size() - 1 && batch.lastOffset > subMap.lastEntry().getValue().lastOffset) {
+                    log.debug("Request batch: {} has offsets which are not found for share partition: {}-{}", batch, groupId, topicIdPartition);
+                    throwable = new InvalidRequestException("Batch record not found");
+                    break;
+                }
+
+                // Add the gap offsets. There is no need to roll back the gap offsets for any
+                // in-flight batch if the acknowledgement request for any batch fails as once
+                // identified gap offset should remain the gap offset in future requests.
+                Set<Long> gapOffsetsSet = null;
+                if (batch.gapOffsets != null && !batch.gapOffsets.isEmpty()) {
+                    // Keep a set to easily check if the gap offset is part of the iterated in-flight
+                    // batch.
+                    gapOffsetsSet = new HashSet<>(batch.gapOffsets);
+                }
+
+                boolean updateNextFetchOffset = batch.acknowledgeType == AcknowledgeType.RELEASE;
+                // The acknowledgement batch either is exact fetch equivalent batch (mostly), subset
+                // or spans over multiple fetched batches. The state can vary per offset itself from
+                // the fetched batch in case of subset.
+                for (Map.Entry<Long, InFlightBatch> entry : subMap.entrySet()) {
+                    InFlightBatch inFlightBatch = entry.getValue();
+
+                    if (!inFlightBatch.memberId().equals(memberId)) {
+                        log.debug("Member {} is not the owner of batch record {} for share partition: {}-{}",
+                            memberId, inFlightBatch, groupId, topicIdPartition);
+                        throwable = new InvalidRequestException("Member is not the owner of batch record");
+                        break;
+                    }
+
+                    boolean fullMatch = inFlightBatch.baseOffset() >= batch.baseOffset
+                        && inFlightBatch.lastOffset() <= batch.lastOffset;
+
+                    if (!fullMatch || inFlightBatch.offsetState != null) {
+                        log.debug("Subset or offset tracked batch record found for acknowledgement,"
```

Comment:
Tiny nit. Could we have a space after the comma please to make the trace slightly more readable.

Reply:
Thanks, missed it.